java/clients/src/main/java/sleeper/clients/QueryWebSocketClient.java

```diff
@@ -77,30 +84,44 @@ public QueryWebSocketClient(
         this.clientSupplier = clientSupplier;
     }
 
-    public void submitQuery(Query query) {
-        Client client = clientSupplier.get();
+    public CompletableFuture<List<String>> submitQuery(Query query) throws InterruptedException {
+        client = clientSupplier.get();
         try {
             Instant startTime = Instant.now();
-            client.startQuery(query);
-            while (!client.hasQueryFinished()) {
-                Thread.sleep(500);
-            }
-            LoggedDuration duration = LoggedDuration.withFullOutput(startTime, Instant.now());
-            long recordsReturned = client.getTotalRecordsReturned();
-            out.println("Query took " + duration + " to return " + recordsReturned + " records");
-        } catch (InterruptedException e) {
-        } finally {
+            return client.startQueryFuture(query)
+                    .whenComplete((records, exception) -> {
+                        try {
+                            client.closeBlocking();
+                        } catch (InterruptedException e) {
+                            Thread.currentThread().interrupt();
+                            throw new RuntimeException(e);
+                        }
+                        LoggedDuration duration = LoggedDuration.withFullOutput(startTime, Instant.now());
+                        long recordsReturned = client.getTotalRecordsReturned();
+                        out.println("Query took " + duration + " to return " + recordsReturned + " records");
+                    });
+        } catch (Exception e) {
             try {
                 client.closeBlocking();
-            } catch (InterruptedException e) {
+            } catch (InterruptedException e2) {
+                throw e2;
             }
+            throw e;
+        }
+    }
+
+    public List<String> getResults(Query query) {
+        if (client == null) {
+            return List.of();
+        } else {
+            return client.getResults(query.getQueryId());
         }
     }
```

Comment:
It doesn't look like this method is really needed any more? It also seems like a problem that this stores all the results from every query ever made by the client in memory.

Reply:
Thanks, i noticed this method wasnt actually used anymore so i've removed it.

I dont think the records stored in-memory is that much of a problem though, as a new WebSocketQueryClient will be created every time you run a query, which will create an empty map of results